File: threetenbp/src/androidTest/java/drewhamilton/androiddatetimeformatters/threetenbp/AndroidDateTimeFormatterTest.java
package drewhamilton.androiddatetimeformatters.threetenbp;

import android.os.Build;
import android.util.Log;
import drewhamilton.androiddatetimeformatters.test.TimeSettingTest;
import org.junit.Test;
import org.threeten.bp.LocalTime;
import org.threeten.bp.format.DateTimeFormatter;

import java.text.ParseException;
import java.util.Date;
import java.util.Locale;

import static org.junit.Assert.assertEquals;

public class AndroidDateTimeFormatterTest extends TimeSettingTest {

    private static final String TAG = AndroidDateTimeFormatterTest.class.getSimpleName();

    private static final int SDK_INT_NULLABLE_TIME_SETTING = 28;

    private static final LocalTime TIME = LocalTime.of(16, 44);
    private static final Date LEGACY_TIME;

    static {
        try {
            LEGACY_TIME = getTimeFormat24InUtc().parse("16:44");
        } catch (ParseException e) {
            throw new RuntimeException(e);
        }
    }

    @Test
    public void ofLocalizedTime_nullSystemSettingUsLocale_uses12HourFormat() {
        if (Build.VERSION.SDK_INT < SDK_INT_NULLABLE_TIME_SETTING) {
            Log.i(TAG, "Time setting is not nullable in API " + Build.VERSION.SDK_INT);
            return;
        }

        setTimeSetting(null);
        setTestLocale(Locale.US);

        DateTimeFormatter formatter = AndroidDateTimeFormatter.ofLocalizedTime(getTestContext());
        String formattedTime = formatter.format(TIME);
        assertEquals(expectedFormattedTime(), formattedTime);
    }

    @Test
    public void ofLocalizedTime_12SystemSettingUsLocale_uses12HourFormat() {
        setTimeSetting("12");
        setTestLocale(Locale.US);

        DateTimeFormatter formatter = AndroidDateTimeFormatter.ofLocalizedTime(getTestContext());
        String formattedTime = formatter.format(TIME);
        assertEquals(expectedFormattedTime(), formattedTime);
    }

    @Test
    public void ofLocalizedTime_24SystemSettingUsLocale_uses24HourFormat() {
        setTimeSetting("24");
        setTestLocale(Locale.US);

        DateTimeFormatter formatter = AndroidDateTimeFormatter.ofLocalizedTime(getTestContext());
        String formattedTime = formatter.format(TIME);
        assertEquals(expectedFormattedTime(), formattedTime);
    }

    @Test
    public void ofLocalizedTime_nullSystemSettingItalyLocale_uses24HourFormat() {
        if (Build.VERSION.SDK_INT < SDK_INT_NULLABLE_TIME_SETTING) {
            Log.i(TAG, "Time setting is not nullable in API " + Build.VERSION.SDK_INT);
            return;
        }

        setTimeSetting(null);
        setTestLocale(Locale.ITALY);

        DateTimeFormatter formatter = AndroidDateTimeFormatter.ofLocalizedTime(getTestContext());
        String formattedTime = formatter.format(TIME);
        assertEquals(expectedFormattedTime(), formattedTime);
    }

    @Test
    public void ofLocalizedTime_12SystemSettingItalyLocale_uses12HourFormat() {
        setTimeSetting("12");
        setTestLocale(Locale.ITALY);

        DateTimeFormatter formatter = AndroidDateTimeFormatter.ofLocalizedTime(getTestContext());
        String formattedTime = formatter.format(TIME);
        assertEquals(expectedFormattedTime(), formattedTime);
    }

    @Test
    public void ofLocalizedTime_24SystemSettingItalyLocale_uses24HourFormat() {
        setTimeSetting("24");
        setTestLocale(Locale.ITALY);

        DateTimeFormatter formatter = AndroidDateTimeFormatter.ofLocalizedTime(getTestContext());
        String formattedTime = formatter.format(TIME);
        assertEquals(expectedFormattedTime(), formattedTime);
    }

    private String expectedFormattedTime() {
        return getAndroidTimeFormatInUtc().format(LEGACY_TIME);
    }
}

File: threetenbp/src/androidTest/java/dev/drewhamilton/androidtime/threetenbp/format/AndroidDateTimeFormatterTest.kt
package dev.drewhamilton.androidtime.threetenbp.format

import android.os.Build
import dev.drewhamilton.androidtime.format.test.TimeSettingTest
import org.junit.Assert.assertEquals
import org.junit.Assume.assumeFalse
import org.junit.Test
import org.threeten.bp.LocalTime
import java.util.Date
import java.util.Locale

class AndroidDateTimeFormatterTest : TimeSettingTest() {

    private val expectedFormattedTime: String
        get() = androidTimeFormatInUtc.format(LEGACY_TIME)

    @Test fun ofLocalizedTime_nullSystemSettingUsLocale_uses12HourFormat() {
        assumeFalse(
            "Time setting is not nullable in API ${Build.VERSION.SDK_INT}",
            Build.VERSION.SDK_INT < SDK_INT_NULLABLE_TIME_SETTING
        )

        systemTimeSetting = null
        testLocale = Locale.US

        val formatter = AndroidDateTimeFormatter.ofLocalizedTime(testContext)
        val formattedTime = formatter.format(TIME)
        assertEquals(expectedFormattedTime, formattedTime)
    }

    @Test fun ofLocalizedTime_12SystemSettingUsLocale_uses12HourFormat() {
        systemTimeSetting = TIME_SETTING_12
        testLocale = Locale.US

        val formatter = AndroidDateTimeFormatter.ofLocalizedTime(testContext)
        val formattedTime = formatter.format(TIME)
        assertEquals(expectedFormattedTime, formattedTime)
    }

    @Test fun ofLocalizedTime_24SystemSettingUsLocale_uses24HourFormat() {
        systemTimeSetting = TIME_SETTING_24
        testLocale = Locale.US

        val formatter = AndroidDateTimeFormatter.ofLocalizedTime(testContext)
        val formattedTime = formatter.format(TIME)
        assertEquals(expectedFormattedTime, formattedTime)
    }

    @Test fun ofLocalizedTime_nullSystemSettingItalyLocale_uses24HourFormat() {
        assumeFalse(
            "Time setting is not nullable in API ${Build.VERSION.SDK_INT}",
            Build.VERSION.SDK_INT < SDK_INT_NULLABLE_TIME_SETTING
        )

        systemTimeSetting = null
        testLocale = Locale.ITALY

        val formatter = AndroidDateTimeFormatter.ofLocalizedTime(testContext)
        val formattedTime = formatter.format(TIME)
        assertEquals(expectedFormattedTime, formattedTime)
    }

    @Test fun ofLocalizedTime_12SystemSettingItalyLocale_uses12HourFormat() {
        systemTimeSetting = TIME_SETTING_12
        testLocale = Locale.ITALY

        val formatter = AndroidDateTimeFormatter.ofLocalizedTime(testContext)
        val formattedTime = formatter.format(TIME)
        assertEquals(expectedFormattedTime, formattedTime)
    }

    @Test fun ofLocalizedTime_24SystemSettingItalyLocale_uses24HourFormat() {
        systemTimeSetting = TIME_SETTING_24
        testLocale = Locale.ITALY

        val formatter = AndroidDateTimeFormatter.ofLocalizedTime(testContext)
        val formattedTime = formatter.format(TIME)
        assertEquals(expectedFormattedTime, formattedTime)
    }

    private companion object {
        private val TIME = LocalTime.of(16, 44)
        private val LEGACY_TIME: Date = TIME_FORMAT_24_IN_UTC.parse("16:44")!!
    }
}
